File: ui/web/frontend/src/components/terminal.ts
import { api } from '../api';

type LogEntry = {
  ts: number;
  source: string;
  text: string;
  seq?: number;        // server-assigned monotonic id; absent on optimistic echoes
  optimistic?: boolean; // client-drew this before the server confirmed
};
type TerminalState = {
  active: boolean;
  cwd: string | null;
  cwd_display: string | null;
  home: string | null;
  venv: string | null;
  git_branch: string | null;
  git_dirty: boolean | null;
  last_active_at: number | null;
  foreground_pid: number | null;
  foreground_cmd: string | null;
  locked_by: 'agent' | 'user' | null;
  shell_pid: number | null;
};

const defaultState = (): TerminalState => ({
  active: false,
  cwd: null,
  cwd_display: null,
  home: null,
  venv: null,
  git_branch: null,
  git_dirty: null,
  last_active_at: null,
  foreground_pid: null,
  foreground_cmd: null,
  locked_by: null,
  shell_pid: null,
});

/**
 * Replace the user's home prefix in an absolute path with ``~``.
 *
 * Prefers ``state.home`` when the backend wrote it (new daemon). Falls
 * back to the well-known Mac (``/Users/<name>``) / Linux
 * (``/home/<name>``) prefix heuristics so the HUD still looks right
 * during an upgrade where the daemon is pre-v2.0.34 but the frontend
 * is fresh. Worst case we leave the absolute path untouched.
 */
function homifyPath(cwd: string, home: string | null): string {
  if (!cwd) return cwd;
  if (home && home !== '~') {
    if (cwd === home) return '~';
    if (cwd.startsWith(home + '/')) return '~' + cwd.slice(home.length);
  }
  // Regex fallback: chomp the first "/Users/<anyone>" or "/home/<anyone>".
  const m = /^\/(Users|home)\/[^/]+(\/|$)/.exec(cwd);
  if (m) {
    const tail = cwd.slice(m[0].length);
    return tail ? '~/' + tail : '~';
  }
  return cwd;
}

function escHtml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}

/**
 * Pinned-at-top Terminal card for the Panel tab.
 *
 * Holds all visual state (log buffer, input draft, scroll position) outside
 * the DOM so panel.ts's wholesale `innerHTML = …` re-render can't nuke
 * in-progress user typing. Consumers embed `renderHtml()` into the panel
 * markup and call `rebind(container)` after innerHTML settles.
 *
 * SSE dispatch: `main.ts` forwards `terminal_log` / `terminal_state` /
 * `terminal_rejected` events via `handleEvent(evt)`. The controller updates
 * its buffers and applies incremental DOM updates when already mounted.
 */
class TerminalController {
  private sessionId: string | null = null;
  private state: TerminalState = defaultState();
  // Single source of truth. Every server-confirmed entry has a monotonic
  // `seq` (assigned by TerminalLogger._append_entry); optimistic echoes
  // carry `optimistic:true` and no seq until the matching SSE confirms.
  private entries: LogEntry[] = [];
  private latestSeq: number = -1;
  private inputDraft: string = '';
  // HUD row default is collapsed: long prompts like
  // "(base) ~/agent_core/butterfly-agent: feature/terminal-panel..."
  // are clipped with ``…`` and the whole row becomes a click target
  // that flips to the expanded view (wraps, shows full text). Kept on
  // the controller so panel re-renders don't reset the toggle.
  private hudExpanded: boolean = false;
  private scrollPinnedToBottom: boolean = true;
  // Distance from the bottom of the scrollable output in pixels. Saved
  // every scroll tick so we can restore the user's viewport across
  // panel re-renders (each render() blows .terminal-output's innerHTML
  // away, which resets scrollTop to 0 — making mid-terminal scrolling
  // jump to the top). Only used when `scrollPinnedToBottom` is false.
  private scrollFromBottom: number = 0;
  private loaded: boolean = false;
  private root: HTMLElement | null = null;

  /** Fetch initial state + log tail for the session. Safe to re-call. */
  async attachSession(id: string): Promise<void> {
    this.sessionId = id;
    this.state = defaultState();
    this.entries = [];
    this.latestSeq = -1;
    this.inputDraft = '';
    this.hudExpanded = false;
    this.scrollPinnedToBottom = true;
    this.scrollFromBottom = 0;
    this.loaded = false;
    try {
      const snap = await api.getTerminal(id, 500);
      if (this.sessionId !== id) return; // session switch raced
      this.state = snap.state;
      this.entries = (snap.log as LogEntry[]).slice();
      // Seed latestSeq from the snapshot so concurrent SSE events that
      // were also captured in the snapshot are dropped by the dedup
      // path instead of rendering twice.
      for (const e of this.entries) {
        if (typeof e.seq === 'number' && e.seq > this.latestSeq) {
          this.latestSeq = e.seq;
        }
      }
      this.loaded = true;
      this.applyStateToChrome();
      this.renderOutputFull();
    } catch (err) {
      console.error('Failed to fetch terminal state:', err);
      this.loaded = true;
      this.applyStateToChrome();
    }
  }

  /** Called by main.ts for each SSE event. No-op for non-terminal types. */
  handleEvent(event: { type?: string; [k: string]: unknown }): void {
    if (event.type === 'terminal_log') {
      const seq = typeof event.seq === 'number' ? (event.seq as number) : undefined;
      // Strict dedup on seq: ignore anything we've already processed
      // (covers the SSE-vs-initial-fetch race). Un-sequenced entries
      // shouldn't happen from the server — if one appears, fall back
      // to append-without-dedup rather than dropping it.
      if (typeof seq === 'number' && seq <= this.latestSeq) return;

      const e: LogEntry = {
        ts: Number(event.ts ?? Date.now() / 1000),
        source: String(event.source ?? 'system'),
        text: String(event.text ?? ''),
        seq,
      };

      // Optimistic-echo reconcile: if we have a pending user_cmd draw
      // (client-side, no seq) whose text matches, upgrade that existing
      // entry in place instead of pushing a duplicate.
      if (e.source === 'user_cmd') {
        const idx = this.entries.findIndex(
          (x) => x.optimistic && x.source === 'user_cmd' && x.text.trim() === e.text.trim()
        );
        if (idx >= 0) {
          this.entries[idx] = { ...this.entries[idx], seq, optimistic: false, ts: e.ts };
          if (typeof seq === 'number') this.latestSeq = seq;
          return;
        }
      }

      this.entries.push(e);
      if (typeof seq === 'number') this.latestSeq = seq;
      this.appendLogLine(e);
    } else if (event.type === 'terminal_state') {
      const s = event.state as Partial<TerminalState> | undefined;
      if (s) {
        this.state = { ...this.state, ...s };
        this.applyStateToChrome();
      }
    } else if (event.type === 'terminal_rejected') {
      this.flashRejected(String(event.reason ?? 'locked'));
    }
  }

  /** HTML the panel embeds. rebind() must follow.
   *
   * Idle / "shell not open" state is surfaced via the input-box
   * placeholder + disabled attribute — no floating overlay in the body.
   */
  renderHtml(): string {
    const statusHtml = this.renderStatusRow();
    const hudHtml = this.renderHudRow();
    return `
      <div class="terminal-container" data-loaded="${this.loaded ? '1' : '0'}">
        <div class="terminal-chrome">
          <div class="terminal-title">Terminal</div>
          ${statusHtml}
        </div>
        <div class="terminal-body">
          <pre class="terminal-output" role="log" aria-live="polite"></pre>
        </div>
        ${hudHtml}
        <div class="terminal-inputbar">
          <span class="terminal-prompt">$</span>
          <input type="text" class="terminal-input"
                 autocomplete="off" autocapitalize="off" spellcheck="false"
                 placeholder="Type a command…" />
          <button type="button" class="terminal-btn-interrupt" title="Send Ctrl-C">⚡</button>
        </div>
      </div>
    `;
  }

  private bindHud(root: HTMLElement): void {
    const hud = root.querySelector('.terminal-hud') as HTMLElement | null;
    if (!hud) return;
    const toggle = () => {
      this.hudExpanded = !this.hudExpanded;
      // Flip attribute in-place — no need for a full re-render, the
      // CSS selector keys off [data-expanded]. Keeps focus + avoids a
      // DOM teardown mid-click.
      hud.setAttribute('data-expanded', this.hudExpanded ? '1' : '0');
      hud.setAttribute(
        'title', this.hudExpanded ? 'Click to collapse' : 'Click to expand'
      );
    };
    hud.addEventListener('click', (e) => {
      // Swallow clicks that are really on the input below — the HUD
      // sits snug to the input bar and iOS mis-targets by a px or two.
      // Safe because .terminal-hud has no interactive children.
      e.stopPropagation();
      toggle();
    });
    hud.addEventListener('keydown', (e) => {
      if (e.key === 'Enter' || e.key === ' ') {
        e.preventDefault();
        toggle();
      }
    });
  }

  /** Reconnect listeners + restore state into the freshly-innerHTML'd DOM. */
  rebind(root: HTMLElement): void {
    this.root = root;
    this.bindHud(root);
    const input = root.querySelector('.terminal-input') as HTMLInputElement | null;
    if (input) {
      input.value = this.inputDraft;
      input.addEventListener('input', () => {
        this.inputDraft = input.value;
      });
      // Chinese IME: Enter that confirms a candidate must not submit.
      // Mirrors chat.ts:1035-1047 — track composition state locally so a
      // mid-composition Enter is treated as a candidate-picker keypress.
      let isComposing = false;
      input.addEventListener('compositionstart', () => { isComposing = true; });
      input.addEventListener('compositionend', () => { isComposing = false; });
      input.addEventListener('keydown', (e) => {
        if (e.key === 'Enter' && !e.shiftKey && !isComposing && !e.isComposing) {
          e.preventDefault();
          this.submitInput();
        }
      });
      this.applyStateToInput(input);
      // Auto-focus when panel opens, unless input is disabled.
      if (!this.inputDisabled()) {
        // Defer so other rebinds in the same tick don't steal focus.
        setTimeout(() => {
          if (document.activeElement !== input) {
            // Only steal focus if the user hasn't clicked something else.
            if (!document.activeElement || document.activeElement === document.body) {
              input.focus();
            }
          }
        }, 0);
      }
    }
    const btn = root.querySelector('.terminal-btn-interrupt') as HTMLButtonElement | null;
    if (btn) {
      btn.addEventListener('click', () => this.sendInterrupt());
      if (this.inputDisabled()) btn.setAttribute('disabled', 'disabled');
    }
    const output = root.querySelector('.terminal-output') as HTMLElement | null;
    if (output) {
      output.addEventListener('scroll', () => {
        // Pin to bottom only if user is near the bottom when they scroll.
        const fromBottom =
          output.scrollHeight - output.scrollTop - output.clientHeight;
        this.scrollPinnedToBottom = fromBottom < 24;
        this.scrollFromBottom = Math.max(0, fromBottom);
      });
    }
    this.renderOutputFull();
    this.applyStateToChrome();
  }

  // ── internals ──────────────────────────────────────────────────────

  private agentLocked(): boolean {
    return this.state.locked_by === 'agent';
  }

  private inputDisabled(): boolean {
    // Disabled when the agent holds the lock OR when the shell has not
    // been opened yet (fix: users couldn't use a shell the agent hadn't
    // started — a typed command would silently queue against a
    // never-alive pty).
    return this.agentLocked() || !this.state.active;
  }

  private async submitInput(): Promise<void> {
    if (!this.sessionId) return;
    if (this.inputDisabled()) return;
    const input = this.root?.querySelector('.terminal-input') as HTMLInputElement | null;
    if (!input) return;
    const raw = input.value;
    if (!raw) return;
    // Optimistically echo the user's command line. Mark optimistic so
    // the matching server-side terminal_log can reconcile this entry in
    // place (no duplicate, no out-of-order render).
    const optimistic: LogEntry = {
      ts: Date.now() / 1000,
      source: 'user_cmd',
      text: raw,
      optimistic: true,
    };
    this.entries.push(optimistic);
    this.appendLogLine(optimistic);
    input.value = '';
    this.inputDraft = '';
    try {
      await api.postTerminalInput(this.sessionId, raw + '\n');
    } catch (err) {
      console.error('Failed to submit terminal input:', err);
      this.flashRejected('send-failed');
    }
  }

  private async sendInterrupt(): Promise<void> {
    if (!this.sessionId || this.inputDisabled()) return;
    try {
      await api.postTerminalInterrupt(this.sessionId);
    } catch (err) {
      console.error('Failed to send terminal interrupt:', err);
    }
  }

  private renderHudRow(): string {
    // Mirror of state.json's env fingerprint rendered as a single
    // terminal-prompt-style line above the input bar. Format:
    //   (<venv>) <path>: <git_branch><* if dirty>
    // git part omitted when not in a repo:
    //   (<venv>) <path>
    // Path is ``~``-shortened by the backend (state.cwd_display) with
    // a client-side fallback for pre-v2.0.34 daemons.
    //
    // Display modes:
    //   collapsed (default) — single line, truncated with ``…``,
    //     click anywhere on the row to expand.
    //   expanded            — wraps across lines, shows full text,
    //     click again to collapse.
    if (!this.state.active) {
      return `<div class="terminal-hud"><span class="terminal-hud-empty">—</span></div>`;
    }
    const venv = this.state.venv || '';
    const cwd = this.state.cwd_display
      || homifyPath(this.state.cwd || '', this.state.home);
    const branch = this.state.git_branch || '';
    const dirty = this.state.git_dirty === true;

    let inner = '';
    if (venv) {
      inner += `<span class="terminal-hud-venv">(${escHtml(venv)})</span>`;
    }
    if (cwd) {
      if (inner) inner += ' ';
      inner += `<span class="terminal-hud-path">${escHtml(cwd)}</span>`;
    }
    if (branch) {
      // "git(main*)" — the wrapper self-identifies the field and the
      // dirty marker sits inside the parens next to the branch name.
      // Two colour zones: the "git(" + ")" shell is pink
      // (.terminal-hud-git-label); the branch + "*" inside is sky
      // blue (.terminal-hud-git-branch).
      const body = `${escHtml(branch)}${dirty ? '*' : ''}`;
      inner += (
        `<span class="terminal-hud-sep">:</span> `
        + `<span class="terminal-hud-git-label">git(</span>`
        + `<span class="terminal-hud-git-branch">${body}</span>`
        + `<span class="terminal-hud-git-label">)</span>`
      );
    }
    if (!inner) {
      return `<div class="terminal-hud"><span class="terminal-hud-empty">—</span></div>`;
    }
    const expandedAttr = this.hudExpanded ? '1' : '0';
    const titleHint = this.hudExpanded ? 'Click to collapse' : 'Click to expand';
    return `
      <div class="terminal-hud" data-expanded="${expandedAttr}" role="button"
           tabindex="0" title="${titleHint}">
        <div class="terminal-hud-line">${inner}</div>
      </div>
    `;
  }

  private renderStatusRow(): string {
    if (!this.loaded) return `<div class="terminal-status">loading…</div>`;
    if (!this.state.active) {
      return `<div class="terminal-status terminal-status-idle">idle</div>`;
    }
    const lock = this.state.locked_by;
    const lockPill = lock === 'agent'
      ? '<span class="terminal-lock-pill lock-agent">🔒 agent using</span>'
      : lock === 'user'
        ? '<span class="terminal-lock-pill lock-user">user typing…</span>'
        : '<span class="terminal-lock-pill lock-free">ready</span>';
    return `<div class="terminal-status">${lockPill}</div>`;
  }

  private applyStateToChrome(): void {
    if (!this.root) return;
    const status = this.root.querySelector('.terminal-status');
    if (status) status.outerHTML = this.renderStatusRow();
    const hud = this.root.querySelector('.terminal-hud');
    if (hud) {
      hud.outerHTML = this.renderHudRow();
      // outerHTML replaces the node → the click listener we attached
      // in bindHud is on the gone-stale element. Re-bind against the
      // fresh one that querySelector will find next.
      this.bindHud(this.root);
    }
    const input = this.root.querySelector('.terminal-input') as HTMLInputElement | null;
    if (input) this.applyStateToInput(input);
    const btn = this.root.querySelector('.terminal-btn-interrupt') as HTMLButtonElement | null;
    if (btn) {
      if (this.inputDisabled()) btn.setAttribute('disabled', 'disabled');
      else btn.removeAttribute('disabled');
    }
  }

  private applyStateToInput(input: HTMLInputElement): void {
    if (this.agentLocked()) {
      input.setAttribute('disabled', 'disabled');
      input.placeholder = 'Agent is using terminal…';
    } else if (!this.state.active) {
      input.setAttribute('disabled', 'disabled');
      input.placeholder = 'Shell not open yet — waiting for the agent to call terminal_create';
    } else {
      input.removeAttribute('disabled');
      input.placeholder = 'Type a command…';
    }
  }

  private renderOutputFull(): void {
    if (!this.root) return;
    const output = this.root.querySelector('.terminal-output') as HTMLElement | null;
    if (!output) return;
    output.innerHTML = this.entries.map(e => this.renderEntry(e)).join('');
    if (this.scrollPinnedToBottom) {
      output.scrollTop = output.scrollHeight;
    } else {
      // Restore the user's scroll position (measured as distance from
      // the bottom) so a mid-terminal viewport survives panel re-renders.
      const target = output.scrollHeight - output.clientHeight - this.scrollFromBottom;
      output.scrollTop = Math.max(0, target);
    }
  }

  private appendLogLine(e: LogEntry): void {
    if (!this.root) return;
    const output = this.root.querySelector('.terminal-output') as HTMLElement | null;
    if (!output) return;
    output.insertAdjacentHTML('beforeend', this.renderEntry(e));
    if (this.scrollPinnedToBottom) output.scrollTop = output.scrollHeight;
    // If the user is scrolled-up, do NOT auto-scroll — honour their position.
  }

  private renderEntry(e: LogEntry): string {
    const src = e.source;
    let cls = 'term-line';
    let prefix = '';
    if (src === 'agent_cmd') {
      cls += ' term-agent-cmd';
      prefix = '<span class="term-prompt-agent">$ </span>';
    } else if (src === 'user_cmd') {
      cls += ' term-user-cmd';
      prefix = '<span class="term-prompt-user">$ </span>';
    } else if (src === 'system') {
      cls += ' term-system';
    } else if (src === 'agent_out') {
      cls += ' term-agent-out';
    } else if (src === 'user_out') {
      cls += ' term-user-out';
    } else {
      cls += ' term-other';
    }
    // Do NOT re-escape whitespace — output already contains \n which the
    // <pre> preserves. Embed text in a span so the prefix stays fixed.
    return `<span class="${cls}">${prefix}<span class="term-text">${escHtml(e.text)}</span></span>`;
  }

  private flashRejected(reason: string): void {
    if (!this.root) return;
    const status = this.root.querySelector('.terminal-status');
    if (!status) return;
    const orig = status.innerHTML;
    status.innerHTML = `<span class="terminal-status-rejected">⚠ ${escHtml(reason)}</span>`;
    setTimeout(() => {
      // Re-render status from current state rather than restoring the
      // snapshot — state may have changed during the flash.
      this.applyStateToChrome();
    }, 1800);
    void orig;
  }
}

export const terminalController = new TerminalController();

File: butterfly/session_engine/terminal.py
"""Terminal session-state + append-only log for the web Terminal panel.

Each session's persistent shell writes two files under
``sessions/<id>/core/terminal/``:

- ``state.json``: liveness snapshot — ``active`` / ``cwd`` /
  ``last_active_at`` / ``foreground_pid`` / ``foreground_cmd`` /
  ``locked_by``. Rewritten in place on every chunk; reflects the NOW.
- ``log.jsonl``: append-only stream of ``{ts, source, text}`` entries. The
  ``source`` tag is one of ``agent_cmd`` / ``agent_out`` / ``user_cmd`` /
  ``user_out`` / ``system``; the web panel replays this to render the
  terminal history.

``TerminalLogger`` is the single writer. PtyShell pushes read chunks via
``on_chunk`` → ``append_output``; the executor stamps command lines via
``append_command`` and the lock state via ``mark_locked``. File writes are
short (≤ a few KB) and O(1); the append-only jsonl is safe to tail from
the HTTP reader side without coordination.
"""
from __future__ import annotations

import json
import os
import threading
import time
from pathlib import Path
from typing import Any, Callable


_DEFAULT_STATE: dict[str, Any] = {
    "active": False,
    "cwd": None,          # absolute path (used by snapshot + restore)
    "cwd_display": None,  # ``~``-shortened form rendered by the HUD
    "home": None,         # $HOME at daemon boot — frontend fallback
    "venv": None,
    "git_branch": None,
    "git_dirty": None,    # None | False | True (None outside a repo)
    "last_active_at": None,
    "foreground_pid": None,
    "foreground_cmd": None,
    "locked_by": None,  # None | "agent" | "user"
    "shell_pid": None,
}


class TerminalLogger:
    """Writer for one session's terminal dir. Thread-/coro-safe via lock."""

    def __init__(
        self,
        terminal_dir: Path,
        event_sink: Callable[[dict], None] | None = None,
    ) -> None:
        self._dir = Path(terminal_dir)
        self._dir.mkdir(parents=True, exist_ok=True)
        self._log_path = self._dir / "log.jsonl"
        self._state_path = self._dir / "state.json"
        self._lock = threading.Lock()
        # Optional: mirror every log/state change onto the SSE event bus so
        # the web frontend gets live updates without tailing log.jsonl.
        self._event_sink = event_sink
        # Ensure files exist so consumers can open them unconditionally.
        if not self._log_path.exists():
            self._log_path.touch()
        if not self._state_path.exists():
            self._write_state_locked(dict(_DEFAULT_STATE))
        # Stamp ``$HOME`` at boot so the frontend can homify paths
        # without needing a backend round-trip. Writes directly instead
        # of via ``_patch_state`` because the init-time stamp must not
        # emit a ``terminal_state`` SSE — integration tests assert the
        # exact events.jsonl sequence per chat/tick and a stray event
        # here rearranges their expected order.
        home_dir = os.environ.get("HOME") or os.path.expanduser("~")
        if home_dir and home_dir != "~":
            try:
                with self._lock:
                    state = self._read_state_locked()
                    if state.get("home") != home_dir:
                        state["home"] = home_dir
                        self._write_state_locked(state)
            except OSError:
                pass
        # Monotonic sequence id stamped on every log entry — makes log.jsonl
        # the single source of truth. SSE events carry the same seq so the
        # frontend can (a) dedupe when a `GET /terminal` snapshot races
        # with in-flight SSE, and (b) order events reliably without
        # relying on timestamps. On startup we seed seq from the existing
        # line count so a restarted daemon continues numbering.
        self._seq: int = self._count_existing_lines()

    def _count_existing_lines(self) -> int:
        try:
            n = 0
            with self._log_path.open("rb") as f:
                for _ in f:
                    n += 1
            return n
        except OSError:
            return 0

    # ── path accessors ─────────────────────────────────────────────────

    @property
    def log_path(self) -> Path:
        return self._log_path

    @property
    def state_path(self) -> Path:
        return self._state_path

    @property
    def directory(self) -> Path:
        return self._dir

    # ── log writes ─────────────────────────────────────────────────────

    def append_output(self, source: str, text: str) -> None:
        """Called by PtyShell.on_chunk for every raw chunk read from the pty.

        `source` is ``"agent_out"`` or ``"user_out"`` — whichever side
        drove the most recent write (PtyShell tracks this via
        ``_current_source``).
        """
        if not text:
            return
        self._append_entry({"ts": time.time(), "source": source, "text": text})

    def append_command(self, source: str, text: str) -> None:
        """Stamp a command line (pre-write). Shell echo is disabled, so
        the command wouldn't show up via `on_chunk`; the executor must
        log it explicitly for the panel to render "$ <cmd>"."""
        self._append_entry({"ts": time.time(), "source": source, "text": text})

    def append_system(self, text: str) -> None:
        """System-generated notice (shell restarted / snapshot restored / etc.)."""
        self._append_entry({"ts": time.time(), "source": "system", "text": text})

    def _append_entry(self, entry: dict) -> None:
        """Stamp `seq`, write to log.jsonl, emit matching SSE event.

        The lock protects both the seq increment and the file write so a
        concurrent call can't interleave seq assignment vs disk order.
        Emit happens outside the lock because the sink hits events.jsonl
        which has its own lock and we don't want to nest.
        """
        with self._lock:
            entry["seq"] = self._seq
            self._seq += 1
            line = json.dumps(entry, ensure_ascii=False) + "\n"
            with open(self._log_path, "a", encoding="utf-8") as f:
                f.write(line)
        self._patch_state({"last_active_at": entry["ts"]})
        self._emit({"type": "terminal_log", **entry})

    # ── state.json ─────────────────────────────────────────────────────

    def read_state(self) -> dict:
        with self._lock:
            return self._read_state_locked()

    def _read_state_locked(self) -> dict:
        try:
            data = json.loads(self._state_path.read_text(encoding="utf-8"))
            if not isinstance(data, dict):
                return dict(_DEFAULT_STATE)
            merged = dict(_DEFAULT_STATE)
            merged.update(data)
            return merged
        except (FileNotFoundError, json.JSONDecodeError, OSError):
            return dict(_DEFAULT_STATE)

    def _write_state_locked(self, state: dict) -> None:
        tmp = self._state_path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8")
        os.replace(tmp, self._state_path)

    def _patch_state(self, patch: dict) -> None:
        with self._lock:
            state = self._read_state_locked()
            state.update(patch)
            self._write_state_locked(state)
        self._emit({"type": "terminal_state", "state": state})

    def _emit(self, event: dict) -> None:
        if self._event_sink is None:
            return
        try:
            self._event_sink(event)
        except Exception:
            pass

    def mark_active(self, active: bool, shell_pid: int | None = None) -> None:
        patch: dict[str, Any] = {"active": active}
        if shell_pid is not None:
            patch["shell_pid"] = shell_pid
        patch["last_active_at"] = time.time()
        self._patch_state(patch)

    def mark_locked(self, by: str | None) -> None:
        """`by` ∈ {None, "agent", "user"}."""
        self._patch_state({"locked_by": by, "last_active_at": time.time()})

    def update_foreground(self, pid: int | None, cmd: str | None) -> None:
        self._patch_state({"foreground_pid": pid, "foreground_cmd": cmd})

    def update_cwd(self, cwd: str | None) -> None:
        self._patch_state({"cwd": cwd})

    def update_env(self, venv: str | None, git_branch: str | None) -> None:
        """Legacy two-field update path. Preferred surface is
        ``update_fingerprint`` which covers the full (venv, cwd,
        cwd_display, git_branch, git_dirty) tuple."""
        self._patch_state({"venv": venv, "git_branch": git_branch})

    def update_fingerprint(
        self,
        *,
        venv: str | None,
        cwd: str | None,
        cwd_display: str | None,
        git_branch: str | None,
        git_dirty: bool | None,
    ) -> None:
        """Write the full env fingerprint in one patch so the frontend
        receives a single ``terminal_state`` SSE event per change (not
        five) — a mid-transition HUD would briefly show, say, the new
        branch against the old venv otherwise."""
        self._patch_state({
            "venv": venv,
            "cwd": cwd,
            "cwd_display": cwd_display,
            "git_branch": git_branch,
            "git_dirty": git_dirty,
        })
